plugin/output/loki/loki.go

```diff
@@ -0,0 +1,360 @@
+
+	// plugin metrics
+	sendErrorMetric prometheus.Counter
+}
+
+func init() {
+	fd.DefaultPluginRegistry.RegisterOutput(&pipeline.PluginStaticInfo{
+		Type:    outPluginType,
+		Factory: Factory,
+	})
+}
+
+func Factory() (pipeline.AnyPlugin, pipeline.AnyConfig) {
+	return &Plugin{}, &Config{}
+}
+
+func (p *Plugin) Start(config pipeline.AnyConfig, params *pipeline.OutputPluginParams) {
+	p.controller = params.Controller
+	p.config = config.(*Config)
+	p.logger = params.Logger
+	p.avgEventSize = params.PipelineSettings.AvgEventSize
+
+	batcherOpts := &pipeline.BatcherOptions{
+		PipelineName:   params.PipelineName,
+		OutputType:     outPluginType,
+		Controller:     p.controller,
+		Workers:        p.config.WorkersCount_,
+		BatchSizeCount: p.config.BatchSize_,
+		BatchSizeBytes: p.config.BatchSizeBytes_,
+		FlushTimeout:   p.config.BatchFlushTimeout_,
+		MetricCtl:      params.MetricCtl,
+	}
+
+	backoffOpts := pipeline.BackoffOpts{
+		MinRetention: p.config.Retention_,
+		Multiplier:   float64(p.config.RetentionExponentMultiplier),
+		AttemptNum:   p.config.Retry,
+	}
+
+	onError := func(err error) {
+		var level zapcore.Level
+		if p.config.FatalOnFailedInsert {
+			level = zapcore.FatalLevel
+		} else {
+			level = zapcore.ErrorLevel
+		}
+
+		p.logger.Desugar().Log(level, "can't send data to loki", zap.Error(err),
+			zap.Int("retries", p.config.Retry))
+	}
+
+	p.httpClient = p.newClient(p.config.RequestTimeout_)
+	p.batcher = pipeline.NewRetriableBatcher(
+		batcherOpts,
+		p.out,
+		backoffOpts,
+		onError,
+	)
+
+	p.batcher.Start(context.Background())
+}
+
+func (p *Plugin) Stop() {
+	p.httpClient.CloseIdleConnections()
+	p.batcher.Stop()
+}
+
+func (p *Plugin) Out(event *pipeline.Event) {
+	p.batcher.Add(event)
+}
+
+func (p *Plugin) out(workerData *pipeline.WorkerData, batch *pipeline.Batch) error {
+	if *workerData == nil {
+		*workerData = &data{
+			outBuf: make([]byte, 0, p.config.BatchSize_*p.avgEventSize),
+		}
+	}
+
+	data := (*workerData).(*data)
+
+	// handle too much memory consumption
+	if cap(data.outBuf) > p.config.BatchSize_*p.avgEventSize {
+		data.outBuf = make([]byte, 0, p.config.BatchSize_*p.avgEventSize)
+	}
+
+	root := insaneJSON.Spawn()
+	outBuf := data.outBuf[:0]
+
+	batch.ForEach(func(event *pipeline.Event) {
+		outBuf = root.MutateToNode(event.Root.Node).Encode(outBuf)
+		_ = root.DecodeString("{}")
+	})
+	insaneJSON.Release(root)
+	data.outBuf = outBuf
+
+	p.logger.Debugf("trying to send: %s", outBuf)
+
+	code, err := p.send(context.Background(), outBuf)
+	if err != nil {
+		p.sendErrorMetric.Inc()
+		p.logger.Errorf("can't send data to Loki address=%s: %v", p.config.Address, err.Error())
+
+		// skip retries for bad request
+		if code == http.StatusBadRequest {
+			return nil
+		}
+	} else {
+		p.logger.Debugf("successfully sent: %s", outBuf)
+	}
+
+	return err
+}
+
+func (p *Plugin) send(ctx context.Context, data []byte) (int, error) {
+	output := map[string]interface{}{
```

Comment:
Hello! Good job but it's better to split up messages and extract fields to labels.

Here is example:

There is sample of config for file.d:

```yaml
pipelines:
  loki:
    settings:
      capacity: 1000
    input:
      type: http
      emulate_mode: "elasticsearch"
      address: ":9200"
    actions:
      - type: parse_es
      - type: json_decode
        field: message
    output:
      type: loki
      address: http://localhost:3100
      # you can suggest your structure of config
      timestamp_field: timestamp
      message_field: message
```

File with input messages (data.json):

```json
{"index": {"_index": "index-main","_type": "span"}}
{"timestamp":"2024-08-01T10:31:55.665748976Z","k8s_container":"container","message":"Started"}
{"timestamp":"2024-08-01T10:35:55.665748976Z","k8s_container":"container","message":"Stopped"}
```

send to file.d

```shell
curl "127.0.0.1:9200/input" -v -H 'Content-Type: application/json' -d @data.json
```

And file.d sends to loki:

```json
{
  "streams": [
    {
      "stream": {
        "k8s_container": "container"
      },
      "values": [
          [ "17225143070000000", "Started" ],
          [ "17225143080000000", "Stopped" ]
      ]
    }
  ]
}
```

Feel free for ask any questions!